service/matching/taskReader.go

```diff
@@ -22,114 +22,156 @@ package matching
 
 import (
 	"context"
-	"fmt"
 	"runtime"
 	"time"
 
+	"github.com/uber/cadence/common/log"
 	"github.com/uber/cadence/common/log/tag"
 	"github.com/uber/cadence/common/metrics"
 	"github.com/uber/cadence/common/persistence"
 )
 
 var epochStartTime = time.Unix(0, 0)
 
-func (c *taskListManagerImpl) deliverBufferTasksForPoll() {
-deliverBufferTasksLoop:
+type (
+	taskReader struct {
+		taskBuffer chan *persistence.TaskInfo // tasks loaded from persistence
+		notifyC    chan struct{}              // Used as signal to notify pump of new tasks
+		tlMgr      *taskListManagerImpl
+		// The cancel objects are to cancel the ratelimiter Wait in dispatchBufferedTasks. The ideal
+		// approach is to use request-scoped contexts and use a unique one for each call to Wait. However
+		// in order to cancel it on shutdown, we need a new goroutine for each call that would wait on
+		// the shutdown channel. To optimize on efficiency, we instead create one and tag it on the struct
+		// so the cancel can be called directly on shutdown.
+		cancelCtx  context.Context
+		cancelFunc context.CancelFunc
+		// separate shutdownC needed for dispatchTasks go routine to allow
+		// getTasksPump to be stopped without stopping dispatchTasks in unit tests
+		dispatcherShutdownC chan struct{}
+	}
+)
+
+func newTaskReader(tlMgr *taskListManagerImpl) *taskReader {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &taskReader{
+		tlMgr:               tlMgr,
+		cancelCtx:           ctx,
+		cancelFunc:          cancel,
+		taskBuffer:          make(chan *persistence.TaskInfo, tlMgr.config.GetTasksBatchSize()-1),
```

Comment:
I know this is from existing code. But do you know why it needs to be -1?

Reply:
my best guess: the deliverBufferedTasks() goroutine always dequeues the head of the buffer and blocks trying to dispatch that task to a poller - so, when there are no pollers, the amount of tasks buffered will be len(taskBuffer) + 1 ; the (-1) was an attempt to offset that additional item - I will add a comment about this